### PR TITLE
Add more conflict types to `ConflictFinder`

### PR DESCRIPTION
**Pull request by grasphoper**
## 📝 Summary

Add more conflict types to `ConflictFinder` as per [#226](https://github.com/flashbots/rbuilder/issues/226)

## A brief overview of the changes:
- I added multiple new mapping entries to `ConflictFinder`, namely 
```rust     
    group_balance_reads: HashMap<Address, Vec<usize>>,
    group_balance_writes: HashMap<Address, Vec<usize>>,
    group_contract_creations: HashMap<Address, Vec<usize>>,
    group_contract_destructions: HashMap<Address, Vec<usize>>,
```
- Added conflicts that can be related to those fields
- Added `combine_groups` - a utility fn to reuse code
- Added `ConflictFinder.add_group_to_index` and `ConflictFinder.remove_group_from_index` to reuse code and improve readability; `index` here means search index, or different member mappings of `ConflictFinder`

---

## ✅ I have completed the following steps:

* [x] Run `make lint`
* [X] Run `make test`
* [X] Added tests (if applicable)

### CODE REVIEW

**Comment by grasphoper** [comment]
For anyone reviewing, please take a look at TODO comments I left and let me know what you think!

**Comment by ZanCorDX** [comment]
The original code had some copy/paste (our bad!) but you may avoid it.
Example:
on remove_group_from_index you can create a fn like:
```
    fn remove_group_from_map<K: std::cmp::Eq + std::hash::Hash + Clone>(
        group_id: usize,
        key_to_group: &mut HashMap<K, Vec<usize>>,
        keys: &Vec<K>,
    ) {
        for key in keys {
            let groups = key_to_group.entry(key.clone()).or_default();
            if let Some(idx) = groups.iter().position(|el| *el == group_id) {
                groups.swap_remove(idx);
            }
        }
    }
```
And just call it 6 times.
The same goes for add (maybe somewhere else?).

**Comment by ZanCorDX** [comment]
About the TODO ideas:
1- We could change HashMap to BTreeMap ensuring that SlotKey's ordering compares Address first. Posible con: BTreeMap could be expensive.
2- Use nested HashMaps. The first level for address and the second level for key.
This way we don't get redundant structures.

**Comment by grasphoper** [comment]
@ZanCorDX thank you for all your suggestions. I implemented them and just pushed updated code! Went with a 2-layer `HashMap` as the "address conflict checking" solution.

**Comment by grasphoper** [comment]
Hi @ZanCorDX . Any more changes you would want to this?

File: crates/rbuilder/src/building/builders/parallel_builder/groups.rs [hunk 1]
```diff
@@ -31,15 +32,69 @@ struct GroupData {
     orders: Vec<SimulatedOrder>,
     reads: Vec<SlotKey>,
     writes: Vec<SlotKey>,
+    balance_reads: Vec<Address>,
+    balance_writes: Vec<Address>,
+    created_contracts: Vec<Address>,
+    destructed_contracts: Vec<Address>,
```
Comment (wjmelements) [inline crates/rbuilder/src/building/builders/parallel_builder/groups.rs]: these two can be simplified to code writes

**Comment by wjmelements** [comment]
It is hard to review code that is force-pushed. Every force-push, the entire changeset must be reviewed again.

**Comment by grasphoper** [comment]
Hi @wjmelements sorry about the force-push. I was rebasing this branch to the latest changes from `upstream/develop`, but git saw them as unique new changes, so I forcepush-removed all the commits from upstream. No changes made to this PR

In the force-push, it shows that the force-push was to [7def2e9](https://github.com/flashbots/rbuilder/commit/7def2e9a71380df49dede16bbc5a4a679294bb75) this commit, which was the one you last reviewed. Does this help?

**Comment by grasphoper** [comment]
@wjmelements

I've not worked with cross-repo PRs like this before. Any advice on how I can rebase onto the `upstream/develop` cleanly?

File: crates/rbuilder/src/building/builders/parallel_builder/groups.rs (hunk 1 above)
Comment (grasphoper) [reply crates/rbuilder/src/building/builders/parallel_builder/groups.rs]: Thank you, combined the 2 under `code_writes` in the latest commit

File: crates/rbuilder/src/building/builders/parallel_builder/groups.rs [hunk 2]
```diff
@@ -68,19 +65,16 @@ fn combine_groups(groups: Vec<GroupData>, removed_group_ids: Vec<usize>) -> Grou
...
 
     GroupData {
         orders,
         reads,
         writes,
         balance_reads,
         balance_writes,
-        created_contracts,
-        destructed_contracts,
+        code_writes,
```
Comment (wjmelements) [inline crates/rbuilder/src/building/builders/parallel_builder/groups.rs:77]: nice

**Review by ZanCorDX — CHANGES_REQUESTED** [review]
Can you fix the lint errors so we can merge?

**Comment by grasphoper** [comment]
@ZanCorDX I've looked at those and not sure how to proceed. They're lint errors from other pieces of code, not touched by my changes. Is it a common thing to happen? 

I can fix them in this PR, but seems like a weird issue. LMK if I should just proceed to fixing them

**Comment by grasphoper** [comment]
@ZanCorDX Actually the linting errors were produced only by the newer rust toolchain and after updating it and merging new commits from upstream, the lint warnings seem to have disappeared. Could you rerun the workflows again to see if all is good now?

**Review by ZanCorDX — APPROVED** [review]

**Pull request merged**